### PR TITLE
[tests] add traffic analysis for Cert_5_2_06_RouterDowngrade.py

### PR DESCRIPTION
**Pull request by hjian2017**
Also add a method in `command.py` to verify a properly formatted address release command message.

### CODE REVIEW

File: tests/scripts/thread-cert/command.py [hunk 1]
```diff
@@ -11,9 +11,19 @@ def check_address_notification(command_msg, source_node, destination_node):
...
+def check_address_release(command_msg, destination_node):
+    """Verify it sent a properly formatted address release command message to destination_node.
+    """
+    command_msg.assertCoapMessageRequestUriPath('/a/ar')
+    command_msg.assertCoapMessageContainsTlv(network_layer.Rloc16)
+
+    destination_rloc = destination_node.get_ip6_address(config.ADDRESS_TYPE.RLOC)
+    assert ipv6.ip_address(destination_rloc) == command_msg.ipv6_packet.ipv6_header.destination_address, "Error: The Destination is wrong." 
+
+    command_msg.assertCoapMessageContainsTlv(network_layer.MacExtendedAddress)
```
Comment (librasungirl) [inline tests/scripts/thread-cert/command.py]: prefer put Line#29 after Line#24.

File: tests/scripts/thread-cert/command.py [hunk 2]
```diff
@@ -11,9 +11,19 @@ def check_address_notification(command_msg, source_node, destination_node):
...
     assert ipv6.ip_address(destination_rloc) == command_msg.ipv6_packet.ipv6_header.destination_address, "Error: The IPv6 Destination address is not the RLOC of the destination."
+
+def check_address_release(command_msg, destination_node):
+    """Verify it sent a properly formatted address release command message to destination_node.
+    """
+    command_msg.assertCoapMessageRequestUriPath('/a/ar')
+    command_msg.assertCoapMessageContainsTlv(network_layer.Rloc16)
+
+    destination_rloc = destination_node.get_ip6_address(config.ADDRESS_TYPE.RLOC)
+    assert ipv6.ip_address(destination_rloc) == command_msg.ipv6_packet.ipv6_header.destination_address, "Error: The Destination is wrong."
```
Comment (librasungirl) [inline tests/scripts/thread-cert/command.py]: Let the error prompts more descriptive?
> "Error: The Destination is not RLOC address"

File: tests/scripts/thread-cert/Makefile.am [hunk 3]
```diff
@@ -261,7 +262,6 @@ TESTS                                                              = \
     $(NULL)
 
 XFAIL_NCP_TESTS                                                    = \
-    Cert_5_2_05_AddressQuery.py                                      \
```
Comment (librasungirl) [inline tests/scripts/thread-cert/Makefile.am]: removed by mistake?

File: tests/scripts/thread-cert/command.py [hunk 4]
```diff
@@ -11,9 +11,18 @@ def check_address_notification(command_msg, source_node, destination_node):
...
 
-    #pdb.set_trace()
     source_rloc = source_node.get_ip6_address(config.ADDRESS_TYPE.RLOC)
     assert ipv6.ip_address(source_rloc) == command_msg.ipv6_packet.ipv6_header.source_address, "Error: The IPv6 Source address is not the RLOC of the originator."
 
     destination_rloc = destination_node.get_ip6_address(config.ADDRESS_TYPE.RLOC)
     assert ipv6.ip_address(destination_rloc) == command_msg.ipv6_packet.ipv6_header.destination_address, "Error: The IPv6 Destination address is not the RLOC of the destination."
+
+def check_address_release(command_msg, destination_node):
+    """Verify it sent a properly formatted address release command message to destination_node.
```
Comment (bukepo) [inline tests/scripts/thread-cert/command.py]: what is `it`? Is it better to say:
```
Verify the message is a properly formatted address release destined to the given node.
```

File: tests/scripts/thread-cert/Cert_5_2_06_RouterDowngrade.py [hunk 5]
```diff
@@ -31,43 +31,72 @@
...
+        dut_messages = self.sniffer.get_messages_sent_by(DUT_ROUTER1)
+        dut_messages.next_mle_message(mle.CommandType.PARENT_REQUEST)
+        dut_messages.next_mle_message(mle.CommandType.CHILD_ID_REQUEST)
+
+        # Verify it sent an Address Release Message to the Leader when it attached as a child.
+        self.assertEqual(self.nodes[DUT_ROUTER1].get_state(), 'child')
+        msg = dut_messages.next_coap_message('0.02')
+        command.check_address_release(msg, self.nodes[LEADER])
+
+        # 4 & 5
```
Comment (bukepo) [inline tests/scripts/thread-cert/Cert_5_2_06_RouterDowngrade.py]: Checking leader response missing?

File: tests/scripts/thread-cert/config.py [hunk 6]
```diff
@@ -51,6 +51,7 @@
                                 0x88, 0x99, 0xaa, 0xbb, 0xcc, 0xdd, 0xee, 0xff])
 
 ADDRESS_TYPE = Enum('ADDRESS_TYPE', ('LINK_LOCAL', 'GLOBAL', 'RLOC', 'ALOC', 'ML_EID'))
+RSSI = {'LINK_QULITY_0': -100, 'LINK_QULITY_1': -95, 'LINK_QULITY_2': -85, 'LINK_QULITY_3': -65}
```
Comment (bukepo) [inline tests/scripts/thread-cert/config.py]: What's this for?

Comment (librasungirl) [reply tests/scripts/thread-cert/config.py]: Enumerating typical RSSI used for later test cases which require special link quality, e.g. 5.3.5

File: tests/scripts/thread-cert/Cert_5_2_06_RouterDowngrade.py (hunk 5 above)
Comment (librasungirl) [reply tests/scripts/thread-cert/Cert_5_2_06_RouterDowngrade.py]: I don't have strong opinion on this as only traffic from DUT are cared in Test Plan.

File: tests/scripts/thread-cert/Makefile.am [hunk 7]
```diff
@@ -122,6 +122,7 @@ EXTRA_DIST                                                         = \
     Cert_9_2_18_RollBackActiveTimestamp.py                           \
     coap.py                                                          \
     common.py                                                        \
+    command.py                                                       \
```
Comment (jwhui) [inline tests/scripts/thread-cert/Makefile.am]: nit: maintain alphabetical order.

File: tests/scripts/thread-cert/Cert_5_2_06_RouterDowngrade.py [hunk 8]
```diff
@@ -31,43 +31,72 @@
...
     def tearDown(self):
+        self.sniffer.stop()
+        del self.sniffer
+
         for node in list(self.nodes.values()):
             node.stop()
         del self.nodes
 
     def test(self):
+        # 1 Ensure topology is formed correctly without the ROUTER24.
```
Comment (jwhui) [inline tests/scripts/thread-cert/Cert_5_2_06_RouterDowngrade.py]: nit: `without the ROUTER24` -> `without ROUTER24`

File: tests/scripts/thread-cert/Cert_5_2_06_RouterDowngrade.py [hunk 9]
```diff
@@ -31,43 +31,72 @@
...
+        for i in range(2, 24):
             self.nodes[i].start()
             time.sleep(5)
             self.assertEqual(self.nodes[i].get_state(), 'router')
 
-        self.nodes[25].start()
+        # This method flushes the message queue so calling this method again will return only the newly logged messages.
+        dut_messages = self.sniffer.get_messages_sent_by(DUT_ROUTER1)
+
+        # 2 ROUTER24: Attach to network. We don't need to ensure it has a better link of quality on posix.
```
Comment (jwhui) [inline tests/scripts/thread-cert/Cert_5_2_06_RouterDowngrade.py]: Why is posix special?

nit: `link of quality` -> `link quality`

Comment (librasungirl) [reply tests/scripts/thread-cert/Cert_5_2_06_RouterDowngrade.py]: Quoting Step 2 of Test Plan here:
> Harness causes Router_24 to attach to the network and ensures it has a link of quality 2 or better to Router_1 and Router_2

No special whitelist or blacklist setting in this case. I think the intention of such requirement is to consider in real test scenarios Router24 may have a poor link connection. However in posix, the link quality should not be an issue. Besides, all reference testbed devices have been configured with downgrade threshold as 32 except the DUT, so only DUT has the chance to downgrade even without setting the link quality.

Thoughts?

File: tests/scripts/thread-cert/Cert_5_2_06_RouterDowngrade.py [hunk 10]
```diff
@@ -31,43 +31,73 @@
...
             self.nodes[i].start()
             time.sleep(5)
             self.assertEqual(self.nodes[i].get_state(), 'router')
 
-        self.nodes[25].start()
+        # This method flushes the message queue so calling this method again will return only the newly logged messages.
+        dut_messages = self.sniffer.get_messages_sent_by(DUT_ROUTER1)
+
+        # 2 ROUTER24: Attach to network.
+        # All reference testbed devices have been configured with downgrade threshold as 32 except DUT_ROUTER1, so we don't need to ensure ROUTER24 has a better link quality on posix.
```
Comment (jwhui) [inline tests/scripts/thread-cert/Cert_5_2_06_RouterDowngrade.py]: nit: limit line length to 120 characters.

**Pull request merged**